### PR TITLE
[@paypal/payouts-sdk] Export types and interfaces

### PR DESCRIPTION
**Pull request by AverageHelper**
- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] [Run `npm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

Changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes:
  - Exported types may be used internally. Exported types are useful for composing values to use as function parameters. ([Source on `export type`](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-8.html#type-only-imports-and-export))
- [x] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header. (Not needed. No new version of the JS library.)

#### TL;DR:
Exporting interface declarations. They're useful.

### CODE REVIEW

**Comment by msrumon** [comment]
@AverageHelper Thanks for the contribution. Those types are for internal use, so that TS compiler can infer the types automatically. Can we refactor this so that we don't need to export internal types and let the compiler infer them? Looking forward to hearing from you.

**Comment by AverageHelper** [comment]
@msrumon Thanks for getting back so quickly! While the compiler _can_ infer these types automatically, internal types are not easy to work with in the case of composing values to pass to PayPal functions.

For example, one of my projects has a function which is designed to return the result of executing a `paypal.payouts.PayoutsItemGetRequest`, after some error handling logic. This project is under a constraint which requires return types to be specified in the signature. Rather than specifying that the function returns a `Promise<ReturnType<typeof paypal.core.PayPalHttpClient.execute>.result>`, I'd rather just import the `GetPayoutsItemResponse` type and declare that my function returns a `Promise<GetPayoutsItemResponse>`.

Another example: another part of my project needs to create a `CreatePayoutRequestBody` to construct a `paypal.payouts.PayoutsPostRequest`. Some of the `PayoutItem`s have properties that we conditionally leave unset. We need to use an `if` statement between declaring the `PayoutItem` and passing that item to the `CreatePayoutRequestBody#items` array. While the compiler _can_ infer the type of this object and check that the object's type is correct at the `request.requestBody` call site, if something does go wrong, that error message is very difficult to read. Furthermore, adding properties conditionally like we do in our `if` statement is not a simple task under TypeScript's type inference. Declaring type conformance beforehand saves us several lines of work.

As far as I know, `export`ing types does not affect type inference where those types are used internally. Is there some other way for external constants to declare conformance to internal types without referencing those internal type declarations?

**Comment by AverageHelper** [comment]
To further illustrate the point of my second example, here is a comparison of the type errors I see with vs. without declaring type conformance early.

Without declaring type conformance, pure compiler inference:
Without type conformance, pure compiler inference

Explicitly declaring type conformance to `CreatePayoutRequestBody`, minimal compiler inference:
Explicitly declaring type conformance to `CreatePayoutRequestBody`, minimal compiler inference

**Comment by msrumon** [comment]
I understand what you're trying to mean. Though I'm not expert in TS type definitions, I'm well-aware that we can make use of the type inference w/o exporting those types. Recently I've worked with [AdonisJS](https://adonisjs.com/) and there the framework nicely infer types quite amazingly. I'm yet to learn how all of these are wired up together.

**Review by msrumon — APPROVED** [review]
Approving for now. But I wish someone would refactor these definitions in future so that compiler can better infer the types.

**Comment by msrumon** [comment]
@AverageHelper Think twice before you merge. I really wished those internal types would never go out in the wild.

**Comment by AverageHelper** [comment]
@msrumon 
Type inference works great when the compiler has sufficient context. For example:

```TypeScript
import * as paypal from "@paypal/payouts-sdk";
import type { CreatePayoutRequestBody, PayoutItem } from "@paypal/payouts-sdk";

const request = new paypal.payouts.PayoutsPostRequest();
request.requestBody({
  sender_batch_header: {
    sender_batch_id: "a_random_id"
  },
  items: []
});
// The compiler understands here what type the object argument is
```

This case works even when the types involved are exported by @paypal/payouts-sdk. The client need not import these types for the compiler to infer them.

However, there are many cases where code authors cannot or do not wish to rely on type inference. For example:

```TypeScript
import * as paypal from "@paypal/payouts-sdk";

const request = new paypal.payouts.PayoutsPostRequest();
const payoutItem: PayoutItem = {
  recipient_type: usingVenmo ? "PHONE" : "EMAIL",
  amount: {
    value: "50.00",
    currency: "USD"
  },
  note: `Thank you for referring ${referral.name}!`,
  sender_item_id: bonusPaymentId,
  receiver
};
if (usingVenmo) {
  // This step requires lots of duplicate code if we were to rely on type inference
  payoutItem.recipient_wallet = "VENMO";
}

const payoutRequest: CreatePayoutRequestBody = {
  sender_batch_header: {
    sender_batch_id: "a_random_id"
  },
  items: [payoutItem]
};

// If we relied on type inference, we could not make this logging call at all. We would not have access to the parameters of the `request.requestBody` function call
logger.log(`${usingVenmo ? "VENMO" : "PAYPAL"} Payout Request`, payoutRequest);
request.requestBody(payoutRequest);
```

**Comment by AverageHelper** [comment]
> I really wished those internal types would never go out in the wild.

Why is this? As far as I know, exporting these types does not affect existing run-time or design-time considerations. Is there some specific concern regarding exported types of which I am not aware?

**Comment by AverageHelper** [comment]
Ready to merge, since I see no pressing issue with exporting type definitions. If these definitions are incorrect, then someone (myself, or somebody somewhere) may correct them in the normal way.
